Add way to mark IPC messages as being deferred if receiving process is suspended

bnham opened the pull request:
#### 9fa5c77109491231f31e5096788292a4722b28cd
<pre>
Add way to mark IPC messages as being deferred if receiving process is suspended
<a href="https://bugs.webkit.org/show_bug.cgi?id=283466">https://bugs.webkit.org/show_bug.cgi?id=283466</a>
<a href="https://rdar.apple.com/137988046">rdar://137988046</a>

Reviewed by Ryan Reno.

This adds a way to mark certain IPC messages as being deferred if the receiving process is
suspended. We are already doing this one-off in several places (e.g. 275588@main and 277427@main)
and there are more IPCs where this might be useful.

To mark a message as being deferred while the process is suspended, use either the option
`DeferSendingIfSuspended` or `DeferSendingIfSuspendedWithCoalescingKeys`. When one of these options
are used, `AuxiliaryProcessProxy::send` will send the message immediately if the receiver is
running, or it will queue the message for sending once the process resumes. The queued message
is coalesced based on the message name, destination ID, and some of the arguments in the message
itself (if the `DeferSendingIfSuspendedWithCoalescingKeys` option is used).

For instance, in this patch:

- We revert the changes in 275588@main and mark `SetHiddenPageDOMTimerThrottlingIncreaseLimit` with the `DeferSendingIfSuspended` option. This means that we&apos;ll keep the latest version that message and send it once the web process resumes.
- We revert the changes in 277427@main and mark `PreferenceDidUpdate` with the `DeferSendingIfSuspendedWithBatchingKeys=(domain, key)` option. This means that we&apos;ll keep the latest version of the PreferenceDidUpdate for a given domain and key and send it once the web process resumes.
- We mark `SetShouldPlayToPlaybackTarget` with the `DeferSendingIfSuspendedWithBatchingKeys=(contextId)` option. This was showing up as the most common message clogging up suspended process message queues in one of our long-running stress tests. I talked to Eric Carlson about this and he said that it&apos;d be safe to coalesce these messages by contextId and send them once the process resumes.

* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
* Source/WebKit/Scripts/webkit/messages_unittest.py:
* Source/WebKit/Scripts/webkit/model.py:
(Message.__init__):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
(parse_enabled_by_string):
(parse_coalescing_keys):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::messageArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOption.messages.in: Copied from Source/WebKit/Shared/AuxiliaryProcess.messages.in.
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOptionMessageReceiver.cpp: Added.
(WebKit::TestWithDeferSendingOption::didReceiveMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithDeferSendingOption_NoOptions&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithDeferSendingOption_NoIndices&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithDeferSendingOption_OneIndex&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithDeferSendingOption_MultipleIndices&gt;):
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOptionMessages.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h.
(Messages::TestWithDeferSendingOption::messageReceiverName):
(Messages::TestWithDeferSendingOption::NoOptions::name):
(Messages::TestWithDeferSendingOption::NoOptions::NoOptions):
(Messages::TestWithDeferSendingOption::NoOptions::arguments):
(Messages::TestWithDeferSendingOption::NoIndices::name):
(Messages::TestWithDeferSendingOption::NoIndices::NoIndices):
(Messages::TestWithDeferSendingOption::NoIndices::encodeCoalescingKey const):
(Messages::TestWithDeferSendingOption::NoIndices::arguments):
(Messages::TestWithDeferSendingOption::OneIndex::name):
(Messages::TestWithDeferSendingOption::OneIndex::OneIndex):
(Messages::TestWithDeferSendingOption::OneIndex::encodeCoalescingKey const):
(Messages::TestWithDeferSendingOption::OneIndex::arguments):
(Messages::TestWithDeferSendingOption::MultipleIndices::name):
(Messages::TestWithDeferSendingOption::MultipleIndices::MultipleIndices):
(Messages::TestWithDeferSendingOption::MultipleIndices::encodeCoalescingKey const):
(Messages::TestWithDeferSendingOption::MultipleIndices::arguments):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamServerConnectionHandleMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassAndWantsAsyncDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassAndWantsDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithWantsAsyncDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchNoSyncMessagesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessages.h:
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/AuxiliaryProcess.messages.in:
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::preferencesDidUpdate): Deleted.
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::sendMessageAfterResuming):
(WebKit::AuxiliaryProcessProxy::didChangeThrottleState):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::send):
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
(WebKit::AuxiliaryProcessProxy::notifyPreferencesChanged):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::updateHiddenPageThrottlingAutoIncreaseLimit):
(WebKit::WebProcessPool::hiddenPageThrottlingAutoIncreaseLimit const): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
(WebKit::WebProcessPool::sendToAllProcesses):
(WebKit::WebProcessPool::sendToAllProcessesForSession):
(WebKit::WebProcessPool::sendToAllRemoteWorkerProcesses):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::sendProcessDidResume):
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/286968@main">https://commits.webkit.org/286968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/641585022bb64c863656141231d0949700e374eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77562 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30478 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82261 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28844 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65747 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4894 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60845 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18759 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50744 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66575 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Running apply-patch; Checked out pull request; Running run-api-tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41115 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/77068 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48146 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27169 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69255 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83553 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4942 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3376 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69011 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 63 flakes 154 failures; Running upload-test-results; 67 flakes 71 failures; Compiled WebKit; 2 flakes 69 failures; Running analyze-layout-tests-results") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5098 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66541 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68272 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17082 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12285 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4889 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7710 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4908 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8343 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6667 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -591,10 +605,28 @@ void AuxiliaryProcessProxy::didChangeThrottleState(ProcessThrottleState state)
+        for (auto& indexMessagePair : m_messagesToSendOnResume.values())
+            indexMessagePairs.append(WTFMove(indexMessagePair));
+
+        // Send messages in the order that they were enqueued after batching.
```
Could use LinkedHashSet with a custom hasher here instead, although this seemed easier to understand to me.

rreno commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -272,6 +273,19 @@ def message_to_struct_declaration(receiver, message):
+        if len(message.coalescing_key_indices) > 0:
+            get_arguments_string = ' << '.join((f'std::get<{i}>(m_arguments)' for i in message.coalescing_key_indices))
+            result.append(f'        encoder << {get_arguments_string};\n')
+        else:
```
nit: Mark the argument with `[[maybe_unused]]`

bnham replied:
I'll just change it so that we explicitly emit `encodeCoalescingKey(IPC::Encoder& encoder)` or `encodeCoalescingKey(IPC::Encoder&)` depending on whether the param is used and get rid of the `UNUSED_PARAM` thing.

rreno replied:
Nice!

rreno commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -591,10 +605,28 @@ void AuxiliaryProcessProxy::didChangeThrottleState(ProcessThrottleState state)
+        indexMessagePairs.reserveInitialCapacity(m_messagesToSendOnResume.size());
+
+        for (auto& indexMessagePair : m_messagesToSendOnResume.values())
+            indexMessagePairs.append(WTFMove(indexMessagePair));
```
Should we set a limit on number of messages to send when resuming a process? Is the coalescing enough to prevent us from spamming the newly-unsuspended process with messages such that it hangs on resume?

bnham replied:
I don't think it's a particular concern for now. We already spam a resuming process with up to 256 messages at once (see `largeOutgoingMessageQueueCountThreshold`--it's used to size two different queues) and given the coalescing here I think we'll be doing much less spamming from this codepath. But if necessary we can add a similar limit here in the future.

bnham committed:
Add way to mark IPC messages as being deferred if receiving process is suspended
https://bugs.webkit.org/show_bug.cgi?id=283466
rdar://137988046

Reviewed by Ryan Reno.

This adds a way to mark certain IPC messages as being deferred if the receiving process is
suspended. We are already doing this one-off in several places (e.g. 275588@main and 277427@main)
and there are more IPCs where this might be useful.

To mark a message as being deferred while the process is suspended, use either the option
`DeferSendingIfSuspended` or `DeferSendingIfSuspendedWithCoalescingKeys`. When one of these options
are used, `AuxiliaryProcessProxy::send` will send the message immediately if the receiver is
running, or it will queue the message for sending once the process resumes. The queued message
is coalesced based on the message name, destination ID, and some of the arguments in the message
itself (if the `DeferSendingIfSuspendedWithCoalescingKeys` option is used).

For instance, in this patch:

- We revert the changes in 275588@main and mark `SetHiddenPageDOMTimerThrottlingIncreaseLimit` with the `DeferSendingIfSuspended` option. This means that we'll keep the latest version that message and send it once the web process resumes.
- We revert the changes in 277427@main and mark `PreferenceDidUpdate` with the `DeferSendingIfSuspendedWithBatchingKeys=(domain, key)` option. This means that we'll keep the latest version of the PreferenceDidUpdate for a given domain and key and send it once the web process resumes.
- We mark `SetShouldPlayToPlaybackTarget` with the `DeferSendingIfSuspendedWithBatchingKeys=(contextId)` option. This was showing up as the most common message clogging up suspended process message queues in one of our long-running stress tests. I talked to Eric Carlson about this and he said that it'd be safe to coalesce these messages by contextId and send them once the process resumes.

* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
* Source/WebKit/Scripts/webkit/messages_unittest.py:
* Source/WebKit/Scripts/webkit/model.py:
(Message.__init__):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
(parse_enabled_by_string):
(parse_coalescing_keys):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::messageArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOption.messages.in: Copied from Source/WebKit/Shared/AuxiliaryProcess.messages.in.
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOptionMessageReceiver.cpp: Added.
(WebKit::TestWithDeferSendingOption::didReceiveMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithDeferSendingOption_NoOptions>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithDeferSendingOption_NoIndices>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithDeferSendingOption_OneIndex>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithDeferSendingOption_MultipleIndices>):
* Source/WebKit/Scripts/webkit/tests/TestWithDeferSendingOptionMessages.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h.
(Messages::TestWithDeferSendingOption::messageReceiverName):
(Messages::TestWithDeferSendingOption::NoOptions::name):
(Messages::TestWithDeferSendingOption::NoOptions::NoOptions):
(Messages::TestWithDeferSendingOption::NoOptions::arguments):
(Messages::TestWithDeferSendingOption::NoIndices::name):
(Messages::TestWithDeferSendingOption::NoIndices::NoIndices):
(Messages::TestWithDeferSendingOption::NoIndices::encodeCoalescingKey const):
(Messages::TestWithDeferSendingOption::NoIndices::arguments):
(Messages::TestWithDeferSendingOption::OneIndex::name):
(Messages::TestWithDeferSendingOption::OneIndex::OneIndex):
(Messages::TestWithDeferSendingOption::OneIndex::encodeCoalescingKey const):
(Messages::TestWithDeferSendingOption::OneIndex::arguments):
(Messages::TestWithDeferSendingOption::MultipleIndices::name):
(Messages::TestWithDeferSendingOption::MultipleIndices::MultipleIndices):
(Messages::TestWithDeferSendingOption::MultipleIndices::encodeCoalescingKey const):
(Messages::TestWithDeferSendingOption::MultipleIndices::arguments):
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledIfMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamServerConnectionHandleMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassAndWantsAsyncDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassAndWantsDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithWantsAsyncDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithWantsDispatchNoSyncMessagesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutUsingIPCConnectionMessages.h:
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/AuxiliaryProcess.messages.in:
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::preferencesDidUpdate): Deleted.
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::sendMessageAfterResuming):
(WebKit::AuxiliaryProcessProxy::didChangeThrottleState):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::send):
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
(WebKit::AuxiliaryProcessProxy::notifyPreferencesChanged):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::updateHiddenPageThrottlingAutoIncreaseLimit):
(WebKit::WebProcessPool::hiddenPageThrottlingAutoIncreaseLimit const): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
(WebKit::WebProcessPool::sendToAllProcesses):
(WebKit::WebProcessPool::sendToAllProcessesForSession):
(WebKit::WebProcessPool::sendToAllRemoteWorkerProcesses):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::sendProcessDidResume):
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: https://commits.webkit.org/286968@main
